### PR TITLE
Fixes #20229 - Fix vlans and bonds deployment

### PR DESCRIPTION
**Pull request by xprazak2**

### CODE REVIEW

**Comment by xprazak2** [comment]
Needs https://github.com/theforeman/community-templates/pull/505 to work properly.

File: app/models/host/base.rb [hunk 1]
```diff
@@ -416,7 +416,7 @@ def get_interface_scope(name, attributes, base = self.interfaces)
 
           mac_based = base.where(:mac => macaddress)
           if attributes[:virtual]
-            mac_based.virtual.where(:identifier => name)
+            mac_based.virtual.where(:identifier => name) || base.where(:attached_to => mac_based.identifier)
```
Comment (ares) [inline app/models/host/base.rb]: Isn't mac_based a collection? I think `.identifier` will fail

also wouldn't it override any interface attached to this identifier?

**Review by ares — COMMENTED** [review]
Question on how it should work

**Comment by xprazak2** [comment]
Correct, `mac_based` is a collection, it wasn't working properly for multiple vlans attached to the same interface, I made a small change.

**Comment by ares** [comment]
Looks reasonable to me now, please only use #20229 as the #24753 tracks other issue too. I'll proceed with testing.

**Comment by xprazak2** [comment]
Reproducer is having 2 vlans on a secondary interface with identifier different from what actually gets deployed.

Before facts upload:







After:

**Comment by lzap** [comment]
@xprazak2 do you think it will solve https://projects.theforeman.org/issues/21184 as well? Edit: Probably not.

**Comment by xprazak2** [comment]
I also think that will have to be fixed separately.

**Comment by xprazak2** [comment]
@lzap, rebased

**Comment by lzap** [comment]
Ok finally getting to this, setting this up. Obviously this needs a test.

File: app/services/puppet_fact_parser.rb [hunk 2]
```diff
@@ -95,10 +95,6 @@ def ipmi_interface
     Hash[ipmi].with_indifferent_access
   end
 
-  # Remove this method override when dropping support for facter < 3.0
```
Comment (ekohl) [inline app/services/puppet_fact_parser.rb:99]: I think this comment is still valid. The whole function isn't used if we detect facter 3.0.

**Comment by lzap** [comment]
So here is how I am trying to reproduce this, @xprazak2 tell me if this is what this is all about.

* Provisioned a simple RHEL7 host, no networking so far.
* Created VLAN interface via `nmcli con add type vlan con-name vlan20 dev eth1 id 20 ip4 10.10.10.10/24`
* This is how network-script generated by NM looks like:
```
VLAN=yes
TYPE=Vlan
PHYSDEV=eth1
VLAN_ID=20
REORDER_HDR=yes
GVRP=no
MVRP=no
PROXY_METHOD=none
BROWSER_ONLY=no
BOOTPROTO=none
IPADDR=10.10.10.10
PREFIX=24
DEFROUTE=yes
IPV4_FAILURE_FATAL=no
IPV6INIT=yes
IPV6_AUTOCONF=yes
IPV6_DEFROUTE=yes
IPV6_FAILURE_FATAL=no
IPV6_ADDR_GEN_MODE=stable-privacy
NAME=vlan20
UUID=d435757e-8a8a-43de-921a-0daaa6b61fcf
ONBOOT=yes
```
* Executed puppet agent to upload facts.
* New (unamanaged) interfaces were created: 
  * **eth1**: #<Nic::Managed id: 18, **mac**: "52:54:00:60:60:02", ip: nil, type: "Nic::Managed", name: "", host_id: 10, subnet_id: nil, domain_id: nil, attrs: {"mtu"=>"1500"}, created_at: "2018-11-09 08:42:33", updated_at: "2018-11-09 08:47:10", provider: nil, username: nil, password: nil, **virtual**: false, link: true, **identifier**: "eth1", tag: "", attached_to: "", **managed**: false, mode: "balance-rr", attached_devices: "", bond_options: "", primary: false, provision: false, compute_attributes: {}, ip6: "", subnet6_id: nil, execution: false>`
  * **eth1.20**: #<Nic::Managed id: 19, **mac**: "52:54:00:60:60:02", ip: "10.10.10.10", type: "Nic::Managed", name: "", host_id: 10, subnet_id: nil, domain_id: nil, attrs: {"netmask"=>"255.255.255.0", "mtu"=>"1500", "network"=>"10.10.10.0"}, created_at: "2018-11-09 08:42:54", updated_at: "2018-11-09 08:47:10", provider: nil, username: nil, password: nil, **virtual**: true, link: true, **identifier**: "eth1.20", **tag**: "20", **attached_to**: "eth1", **managed**: false, mode: "balance-rr", attached_devices: "", bond_options: "", primary: false, provision: false, compute_attributes: {}, ip6: "", subnet6_id: nil, execution: false>
* Edited the host and only marked "eth1.20" as managed
* Rendered provisioning template preview and this is how it looks like:
```
BOOTPROTO="none"
DEVICE=eth1.20
ONBOOT=yes
PEERDNS=no
PEERROUTES=no
DEFROUTE=no
```

Obviously we don't set `VLAN` and `VLAN_ID` parameters in the network-script, is this the bug this PR is all about? I think we ignore `VLAN_ID` on purpose because it is ignored in RHEL6, but I think we can set it since it will be simply ignored (and id will be taken from the interface name).

For the record, I've used facter 2.4 to upload facts, here is the JSON output:

```json
{
  "uptime_seconds": 4309,
  "architecture": "x86_64",
  "virtual": "kvm",
  "is_virtual": true,
  "augeasversion": "1.4.0",
  "vlans": "20",
  "kernel": "Linux",
  "blockdevice_sr0_size": 1073741312,
  "blockdevice_sr0_vendor": "QEMU",
  "blockdevice_sr0_model": "QEMU DVD-ROM",
  "blockdevice_vda_size": 21474836480,
  "blockdevice_vda_vendor": "0x1af4",
  "blockdevices": "sr0,vda",
  "dhcp_servers": {
    "system": "192.168.199.3",
    "eth0": "192.168.199.3"
  },
  "domain": "nat.lan",
  "hardwaremodel": "x86_64",
  "operatingsystem": "CentOS",
  "os": {
    "name": "CentOS",
    "family": "RedHat",
    "release": {
      "major": "7",
      "minor": "5",
      "full": "7.5.1804"
    },
    "lsb": {
      "distcodename": "Core",
      "distid": "CentOS",
      "distdescription": "CentOS Linux release 7.5.1804 (Core) ",
      "release": ":core-4.1-amd64:core-4.1-noarch",
      "distrelease": "7.5.1804",
      "majdistrelease": "7",
      "minordistrelease": "5"
    }
  },
  "facterversion": "2.4.1",
  "filesystems": "xfs",
  "fqdn": "lloyd-albanese.nat.lan",
  "gid": "root",
  "hardwareisa": "x86_64",
  "hostname": "lloyd-albanese",
  "id": "root",
  "interfaces": "eth0,eth1,eth1_20,lo",
  "ipaddress_eth0": "192.168.199.172",
  "macaddress_eth0": "52:54:00:60:60:01",
  "netmask_eth0": "255.255.255.0",
  "mtu_eth0": 1500,
  "macaddress_eth1": "52:54:00:60:60:02",
  "mtu_eth1": 1500,
  "ipaddress_eth1_20": "10.10.10.10",
  "macaddress_eth1_20": "52:54:00:60:60:02",
  "netmask_eth1_20": "255.255.255.0",
  "mtu_eth1_20": 1500,
  "ipaddress_lo": "127.0.0.1",
  "netmask_lo": "255.0.0.0",
  "mtu_lo": 65536,
  "ipaddress": "192.168.199.172",
  "kernelmajversion": "3.10",
  "kernelrelease": "3.10.0-862.14.4.el7.x86_64",
  "kernelversion": "3.10.0",
  "lsbdistcodename": "Core",
  "puppetversion": "3.6.2",
  "lsbdistdescription": "CentOS Linux release 7.5.1804 (Core) ",
  "lsbdistid": "CentOS",
  "lsbdistrelease": "7.5.1804",
  "lsbmajdistrelease": "7",
  "rubyplatform": "x86_64-linux",
  "lsbminordistrelease": "5",
  "lsbrelease": ":core-4.1-amd64:core-4.1-noarch",
  "macaddress": "52:54:00:60:60:01",
  "bios_vendor": "SeaBIOS",
  "bios_version": "?-20180531_142017-buildhw-08.phx2.fedoraproject.org-1.fc28",
  "bios_release_date": "04/01/2014",
  "manufacturer": "QEMU",
  "productname": "Standard PC (i440FX + PIIX, 1996)",
  "serialnumber": "Not Specified",
  "uuid": "5091E293-989E-4E61-8F53-7C6240839293",
  "type": "Other",
  "memorysize": "1.62 GB",
  "memoryfree": "1.39 GB",
  "swapsize": "2.00 GB",
  "swapfree": "2.00 GB",
  "swapsize_mb": "2048.00",
  "swapfree_mb": "2048.00",
  "memorysize_mb": "1655.45",
  "memoryfree_mb": "1423.63",
  "netmask": "255.255.255.0",
  "network_eth0": "192.168.199.0",
  "network_eth1_20": "10.10.10.0",
  "network_lo": "127.0.0.0",
  "uptime_hours": 1,
  "operatingsystemmajrelease": "7",
  "rubysitedir": "/usr/local/share/ruby/site_ruby/",
  "operatingsystemrelease": "7.5.1804",
  "osfamily": "RedHat",
  "partitions": {
    "vda1": {
      "uuid": "0a8a1a49-9bcb-4078-86b1-3e26fb538691",
      "size": "2097152",
      "mount": "/boot",
      "filesystem": "xfs"
    },
    "vda2": {
      "size": "39843840",
      "filesystem": "LVM2_member"
    }
  },
  "path": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin:/sbin",
  "rubyversion": "2.0.0",
  "physicalprocessorcount": 2,
  "processors": {
    "models": [
      "AMD EPYC Processor (with IBPB)",
      "AMD EPYC Processor (with IBPB)"
    ],
    "count": 2,
    "physicalcount": 2
  },
  "processor0": "AMD EPYC Processor (with IBPB)",
  "processor1": "AMD EPYC Processor (with IBPB)",
  "processorcount": 2,
  "ps": "ps -ef",
  "selinux": true,
  "selinux_enforced": true,
  "selinux_policyversion": "31",
  "selinux_current_mode": "enforcing",
  "selinux_config_mode": "enforcing",
  "selinux_config_policy": "unknown",
  "sshrsakey": "AAAAB3NzaC1yc2EAAAADAQABAAABAQDdXrA6PJ6DZHFRR3V0HB/3Nbjz/shN7yNMpzMvzfpJ6m0ss8270lrjTMu55eKnYz57yUlvD1umpXnCWoA9E0aZ2E6pvdjxG3eM0R91M2DvEg94767f2rroDY0zAk79T+wEXTtgIs7UfmpHR9GINatLvo0dOQp98FPrOo6lyWlO2ykGt6UR4DKgI5ShiWys9gSsfNs2oQqdzKnKsWAlmmDKdIVHQnYMVgwugzA/rBhCn9gPMvdyt7ln7DmjBo3/zDhg9aDNULIQa2VXd0F9o3O4glzBP20tcMfXDDgLB4qken4P66fg6LN7yejJrHP96gZE98XwcilyifsQb//FN0dH",
  "sshfp_rsa": "SSHFP 1 1 ffa3e3157f9b1076d8c877f2e874e1c2ad38b5c4\nSSHFP 1 2 b434aea2da324d48ed4fb07d166369ed45f1b19ca956103c94a5a6784cfa7859",
  "sshecdsakey": "AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBCL+rsFM9KsjlxFg33iNbpvEYydCicqxCmX/5qw7sr1/tJEEFrSy+tjgPL+GBOwwQx2QlcWr+Dv2exKFPWt1RgY=",
  "sshfp_ecdsa": "SSHFP 3 1 f38be042d726db424f0ac5c14655e301ac7c13b0\nSSHFP 3 2 52ed18e4bade002233f81152267744ad3f6f5d02a63f876ca047ef4a06d2855e",
  "sshed25519key": "AAAAC3NzaC1lZDI1NTE5AAAAIH1xfdtNmF/pTdL6VXtn5TbW8WRF3bertdq52sKh2C3l",
  "sshfp_ed25519": "SSHFP 4 1 5dafbf1e3dbe521ae7500aaf2a900798f5a492fc\nSSHFP 4 2 753fc5fe86cc533870e77439b8900ff8e4863f634b6e6c4a3872354dee54352e",
  "system_uptime": {
    "seconds": 4309,
    "hours": 1,
    "days": 0,
    "uptime": "1:11 hours"
  },
  "timezone": "UTC",
  "uniqueid": "a8c0acc7",
  "uptime": "1:11 hours",
  "uptime_days": 0
}
```

**Comment by lzap** [comment]
Allright, the missing `VLAN` option appears as soon as I assign subnet. This is probably the correct behavior. I see you have multiple virtual devices, I will try that as well if that's the issue.

**Comment by lzap** [comment]
VLANs work for me, however I deleted the interfaces prior importing new ones. Not sure what I am looking for still.

**Comment by lzap** [comment]
No, I can't get this reproduced.

> Reproducer is having 2 vlans on a secondary interface with identifier different from what actually gets deployed.

Which identifier must be different? I don't see it, probably this is right under my nose on the images but I am unable to find what is actually different.

**Comment by xprazak2** [comment]
Reproducer steps:

1) create a new host in foreman and configure the interfaces as shown on the first screenshot:
* primary and provisioning interface
* secondary interface
* 2 vlans on secondary interface

2) provision the host
3) when the first puppet run is executed on the newly provisioned host, interfaces are incorrectly updated in Foreman.

**Comment by lzap** [comment]
Nah, it works in my environment just fine, the only difference is that I am not using libvirt but bare-metal. I will simply test for regressions and merge.

File: app/services/puppet_fact_parser.rb (hunk 2 above)
Comment (lzap) [reply app/services/puppet_fact_parser.rb:99]: Can you keep this?

File: app/models/host/base.rb [hunk 3]
```diff
@@ -427,6 +427,11 @@ def get_interface_scope(name, attributes, base = self.interfaces)
       end
     end
 
+    def find_by_attached_mac(base, mac_based, identifier, attributes)
+      ifaces = base.where(:attached_to => mac_based.first&.identifier)
+      (ifaces.size > 1) ? ifaces.where(:tag => attributes[:tag]) : ifaces
+    end
```
Comment (lzap) [inline app/models/host/base.rb:433]: Let's create a test that will hit this method.

File: app/services/fact_parser.rb [hunk 4]
```diff
@@ -148,6 +148,8 @@ def set_additional_attributes(attributes, name)
         if @facts[:vlans].present?
           vlans = @facts[:vlans].split(',')
           attributes[:tag] = vlans.include?(tag) ? tag : ''
+        else
+          attributes[:tag] = name.split('.').last
```
Comment (lzap) [inline app/services/fact_parser.rb:152]: And this line.

**Review by lzap — CHANGES_REQUESTED** [review]
Tests are needed for this one. Complicated part that will definitely need rewrite some day in the future.

**Comment by xprazak2** [comment]
I added tests.

**Comment by xprazak2** [comment]
[test upgrade]

**Review by lzap — APPROVED** [review]
I am confident enough with tests, I haven't reproduced this but fact importing works fine. Thanks.

**Pull request merged**